### PR TITLE
Updated auto select gpu

### LINKED ISSUE #1716: Document auto_select_gpus
We forgot to update the docs:
auto_select_gpus

### LINKED ISSUE #2075: Enhancing auto_select_GPU
## 🚀 Feature

Automatically pick a GPU that has enough VRAM for the workload.

### Motivation
Auto_select_GPU #1426 offers potential to automatically adjust the choice of GPU when for example they may be in use by someone else.
However the current implementation only check if it's possible to use the GPU not if the GPU has enough VRAM for the usage.
The proposed feature would have the following workflow : 
- check which GPU can be used
- on train start : 
  - check if one batch can be sent through the network on current GPU
  - else change current GPU
  - repeat until there are no more GPU or a capable GPU has been found
  - (possibly ?) wait a bit then retry

### Pitch
Keep the flag auto_select_GPU but add checks in the lifecycle to ensure you can actually use the GPU selected.

### Alternatives
Add a user try/catch in on_train_start.

### Additional context
The use case for this feature is when you want to run multiple jobs in parallel/on a shared cluster without manually tuning the appropriate GPU.

### PR DESCRIPTION
**Pull request by sebastienwood**
## What does this PR do?

Fixes #2075 (issue)
Fixes #1716

New behavior of auto_select_gpu
- the old behavior is run first, i.e. to check if we can use the GPUs
- the new behavior quick in when the model and dataset are defined
- will possibly pick less GPUs than requested if one of them is not available for the workload
- update constants afterward (`root_gpu`, `on_gpu`, trigger `set_nvidia_flags`)

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

Notes:
- the test does not feel satisfying, looking for ideas to improve it 
- the documentation seems to already convey the effect of this new `auto_select_gpu` behavior, hence no update on that side
- it relies on the definition of the dataloader to work properly, as it is sampled to represent the actual workload
- I have doubts the method is not run too "late" with respect to the env. configuration for the training process
- my first PR on an OSS project with tests and stuff, any comment is welcome !

### CODE REVIEW

File: pytorch_lightning/trainer/distrib_parts.py [hunk 1]
```diff
@@ -471,9 +472,54 @@ def pick_single_gpu(exclude_gpus: list):
...
+        if i in exclude_gpus:
+            continue
+        # Try to allocate on device:
+        device = torch.device(f"cuda:{i}")
+        try:
+            with torch.set_grad_enabled(True):
+                model_device = model.to(device)
+                batch_device = batch.to(device)
+                model_device.train()  # record grads
+                model_device(batch_device)
```
Comment (justusschock) [inline pytorch_lightning/trainer/distrib_parts.py:487]: this probably won't work, since usually a batch here is not only the input tensor, but can be of any type (including dicts and strings), which you don't want to feed into the network

**Comment by SpontaneousDuck** [comment]
A little hesitant about it automatically selecting fewer GPUs than requested if not enough are usable, even if it warns the user. Might be useful to have a flag that would allow the user to choose if this throws an error or automatically reduces number of GPUs.

**Comment by justusschock** [comment]
@SpontaneousDuck actually I strongly disagree here. We have already so many flags, that I don't want to have another one. But I agree that the correct behaviour should be raising an error

**Comment by sebastienwood** [comment]
> The current implementation won't work due to the specified reasons. Another question: Is this called before DP/DDP initialization? Otherwise it won't work with those

It is called line 975 in trainer.py, before the whole if/elif of accelerator backends which setup DP/DDP. 
Would `model_device.training_step(batch_device, 0)` be ok for line 487 ? It should delegate batch handling to the user.

**Comment by SpontaneousDuck** [comment]
@justusschock Sounds great to me! Just was a little worried about reproducibility since a variable number of GPUs can affect that. Just throwing an error instead would solve that. I do agree there are already a ton of flags!

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -1407,6 +1412,15 @@ def call_setup_hook(self, model):
         self.setup(stage_name)
         model.setup(stage_name)
 
+    def update_auto_selected_gpus(self, model):
+        # Called when model/data is known. Ensure the GPU used have enough VRAM.
+        self.gpus = pick_multiple_gpus(len(self.gpus), model)  # At most the current number of GPUs
+
+        self.data_parallel_device_ids = _parse_gpu_ids(self.gpus)
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/trainer.py:1419]: Could this function be used during `__init__` to not have duplicate code?

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1407,6 +1412,15 @@ def call_setup_hook(self, model):
         self.setup(stage_name)
         model.setup(stage_name)
 
+    def update_auto_selected_gpus(self, model):
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/trainer.py:1415]:
```suggestion
    def update_auto_selected_gpus(self, model: LightningModule):
```

**Comment by justusschock** [comment]
@sebastienwood While this PR seems to be good, I have another question:

Let's say I have a system with two GPUs, which are both empty at the beginning and can both hold 10GB.

My Model (including inputs, activations and gradients) takes only 4GB. So When I now tell lightning to search for a free GPU, it will use the first one (which is great). But if I want to run the same experiment in parallel with another config, it would probably select the same GPU again, since there is still memory left. 

However, if my model is computationally expensive, that the GPU utilisation is always high even though I don't need much memory, it will clearly slow down everything. Can't we use free GPUs if available or maybe go for utilisation first and then as a second order criteria for memory?

**Comment by williamFalcon** [comment]
yeah, i agree with @justusschock. We don’t want to add multiple models for the same GPU. for our purposes we should consider any GPU that is being utilized as “full” (even if the memory is only 5% for example)

**Comment by justusschock** [comment]
@williamFalcon actually, we shouldn't be that strict. We should search for empty GPUs, but if no empty gpu is there, you should use the ones, that actually fit from memory perspective, since on workstations you often also have the x-server allocating some GPU Memory (100-200 MB typically)

**Comment by williamFalcon** [comment]
yeah, but the problem is that if you are sharing this with someone you’ll put your work on their gpu and cut their speed in half

**Comment by justusschock** [comment]
That's why I wanted to look at GPU-Utilization instead :)

**Comment by awaelchli** [comment]
But here a compromise/proposition: Implement the strict version in such a way that it can easily be overridden by anyone who wants different behavior or has special requirements on the type of gpu etc.
Instead of a function, make it a class that implements the algorithm. Break it down into the fundamental blocks. Let the user override parts of it or the whole algorithm. Then: 
```python
trainer = Trainer(gpus="auto") # default algorithm
trainer = Trainer(gpus=SophisticatedGPUChooser())  # custom algorithm
```

**Comment by sebastienwood** [comment]
@awaelchli this is appealing. There are a limited number of scenarios (based on the discussions):
- less than the request number of GPU could accept the workload, throw an error
- enough capable GPUs were found, but they are already in use : go/nogo decision
- enough capable GPUs were found, they're free

The only block to clear is the go/nogo decision. Is it ok to have a simple API for the decision, based on the current maximal usage ratio of the VRAM for the capable GPUs ? 

E.g. we have one GPU, we request one. Its VRAM load is 25%. It is capable. `criterion_function(0.25) -> bool`.

**Comment by edenlightning** [comment]
mind rebasing?

**Comment by edenlightning** [comment]
@williamFalcon @awaelchli what do you think of the latest suggestion?

**Comment by awaelchli** [comment]
If a GPU is in use, we should not pick it as a candidate. Even if the memory allows it, you can't really judge the GPU utilisation, and I have not made good experience, with processes crashing. I think it should be one process per gpu max.

**Comment by HanSilver01** [comment]
![Cargando 98921748...]

**Comment by qsh-zh** [comment]
@awaelchli Your suggestion is appealing. I am new to PyTorch lightning, do we have such a feature that we can select the best available device?
